### PR TITLE
Assign a Network Address to a Target

### PR DESCRIPTION
**Pull request by hugoamvieira**
### Summary:

This branch contains a set of commits that enable a Boundary Target to be directly configured with an address to create a session against, instead of having to create Host Catalogs, Host Sources and Hosts.

### How to use (CLI):

Creating a tcp target with an address:
```
boundary targets create tcp -scope-id p_1234567890 -name test -default-port 22 -address "localhost"
```

Updating a tcp target with an address:
```
boundary targets update tcp -id ttcp_1234567890 -address "localhost"
```

Removing an address from a target:
```
boundary targets update tcp -id ttcp_1234567890 -address "null"
```

### Notes:

- The address field can be a dns name or an ipv4 endpoint
- When deleting an address from a target, all active sessions using the target will be canceled. Changing the value of the address on a given target will not cancel any sessions.
- When adding or setting a host source to a target that has a network address directly configured to it, an error should be returned
- When setting a network address directly to a target that has a host source association, an error should be returned

Example of the error:
```
Error from controller when performing add-host-sources on target

Error information:
  Kind:                FailedPrecondition
  Message:             target.(Repository).AddTargetHostSources: error creating sets: db.DoTx:
  target.(Repository).AddTargetHostSources: unable to add host sources because a network address is directly assigned to the
  given target: integrity violation: error #409
  Status:              400
  context:             Error from controller when performing add-host-sources on target
```
**Note**: grpc-gateway [maps](https://github.com/grpc-ecosystem/grpc-gateway/blob/257af994261d2395689810c48b3a3fb8ad7b318a/runtime/errors.go#L36) Failed Precondition errors into http bad request errors because `this deliberately doesn't translate to the similarly named '412 Precondition Failed' HTTP response status.`

### Concerns:

- database migration test 59_01 was deleted because of the breaking changes implemented into the domain code. which includes interacting with the target_address table. This test invoked functions in the domain code, such as LookupTarget & UpdateTarget, which now interacts with the target_address table. Because the target_address table does not exist until migration id 70, a postgress error is returned notifying the user that the table does not exist.

**Important**: We also have [Terraform changes](https://github.com/hashicorp/terraform-provider-boundary/pull/308) to implement client support for a Target using an address. Merging and releasing a provider with those changes is blocked until a Boundary version with the changes in this PR is released. After that, we can update the module imports on the [PR](https://github.com/hashicorp/terraform-provider-boundary/pull/308) and create a new Terraform provider release there.

### CODE REVIEW

File: internal/cmd/commands/targetscmd/ssh_funcs.go [hunk 1]
```diff
@@ -151,6 +158,14 @@ func extraSshFlagsHandlingFuncImpl(c *SshCommand, _ *base.FlagSets, opts *[]targ
 		*opts = append(*opts, targets.WithWorkerFilter(c.flagWorkerFilter))
 	}
 
+	switch c.flagAddress {
+	case "":
+	case "null":
+		*opts = append(*opts, targets.DefaultAddress())
```
Comment (johanbrandhorst) [inline internal/cmd/commands/targetscmd/ssh_funcs.go:164]: I thought we discussed not having this magic `"null"` and instead have the setter and remover calls? If this is the `target create` endpoint, I would expect any presence of the address flag to be a valid address. Maybe they want to connect to a host with a hostname of `null`?

File: internal/proto/controller/api/resources/targets/v1/target.proto [hunk 2]
```diff
@@ -188,6 +188,9 @@ message Target {
   // Output only. The available actions on this resource for this user.
   repeated string authorized_actions = 300 [json_name = "authorized_actions"]; // @gotags: `class:"public"`
 
+  // Optionally, a valid network address to connect to for this target. Can not be used alongside host sources.
+  string address = 540 [(custom_options.v1.generate_sdk_option) = true]; // @gotags: `class:"public"`
```
Comment (johanbrandhorst) [inline internal/proto/controller/api/resources/targets/v1/target.proto:192]: Where does this tag number come from 😁? Even for our convention it seems high.

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Should this be targeting the release branch? I assume not.

**Comment by hugoamvieira** [comment]
> Should this be targeting the release branch? I assume not.

From what I know, that's how we're doing things now. We're not targeting `main` anymore, instead targeting release branches and then cherry-picking the commits into `main` when the PR is merged

File: internal/proto/controller/api/resources/targets/v1/target.proto (hunk 2 above)
Comment (johanbrandhorst) [reply internal/proto/controller/api/resources/targets/v1/target.proto:192]: Huh, ok!

File: internal/cmd/commands/targetscmd/ssh_funcs.go (hunk 1 above)
Comment (hugoamvieira) [reply internal/cmd/commands/targetscmd/ssh_funcs.go:164]: We also mentioned that any address we accept when creating a Host should be what we accept here. Currently, [we don't accept a `null` host address](https://github.com/hashicorp/boundary/blob/main/internal/cmd/commands/hostscmd/static_funcs.go#L77-L83), so this maintains consistency from a CLI perspective.

**Comment by johanbrandhorst** [comment]
> > Should this be targeting the release branch? I assume not.
> 
> From what I know, that's how we're doing things now. We're not targeting `main` anymore, instead targeting release branches and then cherry-picking the commits into `main` when the PR is merged

Uh, I thought that was the case until we were doing a release, but I suppose you could be onto something, good point!

**Comment by johanbrandhorst** [comment]
Lets just make sure we don't merge this until 0.11.1 is out the door 😁

File: internal/cmd/commands/targetscmd/ssh_funcs.go (hunk 1 above)
Comment (johanbrandhorst) [reply internal/cmd/commands/targetscmd/ssh_funcs.go:164]: That's fair, though I hate it :P

**Review by ddebko — CHANGES_REQUESTED** [review]
Need to update the changelog

File: internal/session/testing.go [hunk 3]
```diff
@@ -113,13 +135,61 @@ func TestSession(t testing.TB, conn *db.DB, rootWrapper wrapping.Wrapper, c Comp
...
+	}
+
 	ss, err := fetchStates(ctx, rw, s.PublicId, append(opts.withDbOpts, db.WithOrder("start_time desc"))...)
 	require.NoError(err)
 	s.States = ss
 
 	return s
 }
 
+func TestSessionStaticAddress(t testing.TB, conn *db.DB, wrapper wrapping.Wrapper, iamRepo *iam.Repository, opt ...Option) *Session {
```
[the remark this answers is not in the document]
Comment (ddebko) [reply internal/session/testing.go:151]: @hugoamvieira what do you think of this change?

File: internal/session/testing.go [hunk 4]
```diff
@@ -113,13 +135,61 @@ func TestSession(t testing.TB, conn *db.DB, rootWrapper wrapping.Wrapper, c Comp
...
+	t.Helper()
+	composedOf := TestSessionStaticAddressParams(t, conn, wrapper, iamRepo)
+	future := timestamppb.New(time.Now().Add(time.Hour))
+	exp := &timestamp.Timestamp{Timestamp: future}
+	return TestSession(t, conn, wrapper, composedOf, append(opt, WithExpirationTime(exp))...)
+}
+
+func TestSessionStaticAddressParams(t testing.TB, conn *db.DB, wrapper wrapping.Wrapper, iamRepo *iam.Repository) ComposedOf {
```
[the remark this answers is not in the document]
Comment (ddebko) [reply internal/session/testing.go:159]: @hugoamvieira what do you think of this change?

File: internal/session/repository_session.go [hunk 5]
```diff
@@ -682,6 +712,15 @@ func fetchConnections(ctx context.Context, r db.Reader, sessionId string, opt ..
 	return connections, nil
 }
 
+func fetchHostSetHost(ctx context.Context, r db.Reader, sessionId string, opt ...db.Option) (*SessionHostSetHost, error) {
+	const op = "session.fetchHostSetHost"
+	var hostSetHost *SessionHostSetHost
+	if err := r.SearchWhere(ctx, &hostSetHost, "session_id = ?", []interface{}{sessionId}, opt...); err != nil {
```
Comment (johanbrandhorst) [inline internal/session/repository_session.go:718]: Replace `interface{}` with `any`.
```suggestion
	if err := r.SearchWhere(ctx, &hostSetHost, "session_id = ?", []any{sessionId}, opt...); err != nil {
```

File: internal/tests/api/targets/target_test.go [hunk 6]
```diff
@@ -379,6 +380,166 @@ func TestList(t *testing.T) {
...
+	require.NotNil(t, h)
+	hUpdate, err := hostsets.NewClient(client).AddHosts(tc.Context(), hs.Item.Id, hs.Item.Version, []string{h.GetItem().Id})
+	require.NoError(t, err)
+	require.NotNil(t, hUpdate)
+
+	// Expect error when associating a host source to the target with a direct network address association
+	targetId := targetResp.GetItem().Id
+	version := targetResp.GetItem().Version
+	updateResp, err := tClient.AddHostSources(tc.Context(), targetId, version, []string{hs.Item.Id})
+	require.Error(t, err)
```
Comment (johanbrandhorst) [inline internal/tests/api/targets/target_test.go:418]: Can we verify that this is the expected `409 Conflict` error?

File: internal/tests/api/targets/target_test.go [hunk 7]
```diff
@@ -379,6 +380,166 @@ func TestList(t *testing.T) {
...
+	require.Empty(t, targetResp.GetItem().Address)
+	version = targetResp.GetItem().Version
+	updateResp, err = tClient.AddHostSources(tc.Context(), targetId, version, []string{hs.Item.Id})
+	require.NoError(t, err)
+	require.NotNil(t, updateResp)
+	require.Empty(t, updateResp.GetItem().Address)
+	require.Equal(t, []string{hs.Item.Id}, updateResp.GetItem().HostSourceIds)
+}
+
+func TestTarget_HostSourceMutualExlcusiveRelationShip(t *testing.T) {
```
Comment (johanbrandhorst) [inline internal/tests/api/targets/target_test.go:434]:
```suggestion
func TestTarget_HostSourceMutualExclusiveRelationship(t *testing.T) {
```

File: internal/tests/api/targets/target_test.go [hunk 8]
```diff
@@ -379,6 +380,166 @@ func TestList(t *testing.T) {
...
+	targetId := targetResp.GetItem().Id
+	version := targetResp.GetItem().Version
+	updateResp, err := tClient.AddHostSources(tc.Context(), targetId, version, []string{hs.Item.Id})
+	require.NoError(t, err)
+	require.NotNil(t, updateResp)
+	require.Empty(t, updateResp.GetItem().Address)
+	require.Equal(t, []string{hs.Item.Id}, updateResp.GetItem().HostSourceIds)
+	version = updateResp.GetItem().Version
+	updateResp, err = tClient.Update(tc.Context(), targetId, version, targets.WithAddress("localhost"))
+	require.Error(t, err)
```
Comment (johanbrandhorst) [inline internal/tests/api/targets/target_test.go:475]: Ditto, check the error type

File: internal/tests/cli/boundary/target.bats [hunk 9]
```diff
@@ -100,3 +100,36 @@ load _helpers
...
+  [ "$status" -eq 1 ]
+}
+
+@test "boundary/target: can assign an an address to a target after deleting host source" {
+  local id=$(target_id_from_name $DEFAULT_P_ID $TGT_NAME_WITH_ADDR)
+  run remove_host_sources $id $DEFAULT_HOST_SET
+  [ "$status" -eq 0 ]
+  run update_address $id "localhost"
+  [ "$status" -eq 0 ]
+}
```
Comment (johanbrandhorst) [inline internal/tests/cli/boundary/target.bats:135]: Can we add a test that tries to unset an address by setting it to "null", as we have a handler for in the CLI? And that the host source association can then be set.

File: testing/internal/e2e/tests/static/target_address_test.go [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+	// Remove target address. We have now no address or host sources.
+	output = e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"targets", "update", "tcp",
+			"-id", newTargetId,
+			"-address", "null",
+			"-format", "json",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
```
Comment (johanbrandhorst) [inline testing/internal/e2e/tests/static/target_address_test.go:66]: Nice! 👌🏻

File: testing/internal/e2e/tests/static/target_address_test.go [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+			"-id", newTargetId,
+			"-host-source", newHostSetId,
+			"-format", "json",
+		),
+	)
+	require.Error(t, output.Err, "Target add-host-sources did not return expected error")
+
+	var response boundary.CliError
+	require.NoError(t, json.Unmarshal(output.Stderr, &response))
+	require.Equal(t, http.StatusBadRequest, response.Status, "Expected error when adding host source to target with an address")
```
Comment (johanbrandhorst) [inline testing/internal/e2e/tests/static/target_address_test.go:189]: I thought this was supposed to be `http.Conflict`? I don't hate `BadRequest` but just want to confirm.

File: testing/internal/e2e/tests/static/target_address_test.go [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+			"-o", "UserKnownHostsFile=/dev/null",
+			"-o", "StrictHostKeyChecking=no",
+			"-o", "IdentitiesOnly=yes", // forces the use of the provided key
+			"-p", "{{boundary.port}}", // this is provided by boundary
+			"{{boundary.ip}}",
+			"hostname", "-i",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+}
```
Comment (johanbrandhorst) [inline testing/internal/e2e/tests/static/target_address_test.go:340]: These end-to-end tests are _excellent_. Thank you!

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Looks like the CHANGELOG needs an update too 😄

File: internal/session/testing.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/session/testing.go:151]: I agree - Good spot!

[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/session/testing.go]: [4eb6da0](https://github.com/hashicorp/boundary/pull/2613/commits/4eb6da0cbe05def7b4afde2ad97d78ad02b5b9fc)

File: internal/session/testing.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/session/testing.go]: [4eb6da0](https://github.com/hashicorp/boundary/pull/2613/commits/4eb6da0cbe05def7b4afde2ad97d78ad02b5b9fc)

File: testing/internal/e2e/tests/static/target_address_test.go (hunk 11 above)
Comment (hugoamvieira) [reply testing/internal/e2e/tests/static/target_address_test.go:189]: Yep, this is correct - @ddebko and I thought `BadRequest` made more sense here.

File: internal/db/schema/migrations/oss/postgres_59_01_test.go [hunk 13]
```diff
@@ -1,188 +0,0 @@
...
-	// Attempt to create a new target with all 3 filters- expect failure
-	tar3, err := targettest.New(proj.PublicId, target.WithName("new-worker-filters"), target.WithEgressWorkerFilter(oldWorkerFilter),
-		target.WithIngressWorkerFilter(oldWorkerFilter), target.WithEgressWorkerFilter(oldWorkerFilter))
-	require.NoError(t, err)
-	id, err = db.NewPublicId("ttcp")
-	require.NoError(t, err)
-	tar3.SetPublicId(ctx, id)
-	err = rw.Create(context.Background(), tar3)
-	require.Error(t, err)
-}
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/db/schema/migrations/oss/postgres_59_01_test.go:188]: Nope - This is intentional. Refer to `# Concerns` in the PR description, as well as [this commit](https://github.com/hashicorp/boundary/pull/2613/commits/bba2b9e31883973243e2429cae2c2e64a71d2261).

**Comment by hugoamvieira** [comment]
CHANGELOG entry added with [611f661](https://github.com/hashicorp/boundary/pull/2613/commits/611f6615d605c2b191c3da8ea1594218714feb45)

File: internal/errors/is.go [hunk 14]
```diff
@@ -94,3 +94,17 @@ func IsNotFoundError(err error) bool {
...
+func IsConflictError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if Match(T(Conflict), err) {
+		return true
+	}
+
+	return false
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/errors/is.go]: Yep, good spot [9155463](https://github.com/hashicorp/boundary/pull/2613/commits/91554638fb2b7846db6aa02d431b25c3ed0beea0)

File: internal/daemon/controller/handlers/errors.go [hunk 15]
```diff
@@ -150,6 +150,22 @@ func InvalidArgumentErrorf(msg string, fields map[string]string) *ApiError {
...
+// request because this is how the grpc-gateway mapping is implemented for
+// failed precondition protobuf errors.
+func ConflictErrorf(msg string) *ApiError {
+	const op = "handlers.ConflictErrorf"
+	ctx := context.TODO()
+	err := ApiErrorWithCodeAndMessage(codes.FailedPrecondition, msg)
+	var apiErr *ApiError
+	if !errors.As(err, &apiErr) {
+		event.WriteError(ctx, op, err, event.WithInfoMsg("Unable to build conflict api error."))
+	}
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/daemon/controller/handlers/errors.go:165]: We don't seem to do anything with `apiErr` other than returning it, so the if-statement seems purely informational here. It follows the pattern from existing functions, though.

File: internal/daemon/controller/handlers/targets/target_service.go [hunk 16]
```diff
@@ -1164,8 +1181,14 @@ func (s Service) addHostSourcesInRepo(ctx context.Context, targetId string, host
 	}
 	out, hs, cl, err := repo.AddTargetHostSources(ctx, targetId, version, strutil.RemoveDuplicates(hostSourceIds, false))
 	if err != nil {
+		var internalErr *errors.Err
+		if stderrors.As(err, &internalErr) && internalErr.Code == errors.Conflict {
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/daemon/controller/handlers/targets/target_service.go:1185]: > Do we have an integration test for this case?

I don't think so, we verify this over in the end-to-end tests. Although with https://github.com/hashicorp/boundary/pull/2613#discussion_r1061047106 we'll add the check to the integration tests.

> Also, is this case documented in the protobuf comments which makes its way to our generated API documentation?

Good catch - We only had it in other places - Added with [0bc826e](https://github.com/hashicorp/boundary/pull/2613/commits/0bc826e31b58cd314ec4dca871371c11aabc4c89)

File: internal/tests/api/targets/target_test.go (hunk 7 above)
Comment (hugoamvieira) [reply internal/tests/api/targets/target_test.go]: Fixed [089532f](https://github.com/hashicorp/boundary/pull/2613/commits/089532f36b312f6679813a6dce84c2f57c108a9e)

File: internal/session/repository_session.go (hunk 5 above)
Comment (hugoamvieira) [reply internal/session/repository_session.go]: Fixed [089532f](https://github.com/hashicorp/boundary/pull/2613/commits/089532f36b312f6679813a6dce84c2f57c108a9e)

File: internal/daemon/controller/handlers/targets/target_service.go [hunk 17]
```diff
@@ -1522,7 +1554,20 @@ func validateCreateRequest(req *pbs.CreateTargetRequest) error {
...
+		if address := req.GetItem().GetAddress(); address != nil {
+			if len(address.GetValue()) < static.MinHostAddressLength ||
+				len(address.GetValue()) > static.MaxHostAddressLength {
+				badFields[globals.AddressField] = fmt.Sprintf("Address length must be between %d and %d characters.", static.MinHostAddressLength, static.MaxHostAddressLength)
+			}
+			_, _, err := net.SplitHostPort(address.GetValue())
+			switch {
+			case err == nil:
+				badFields[globals.AddressField] = "Address does not support a port."
+			case strings.Contains(err.Error(), "missing port in address"):
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/daemon/controller/handlers/targets/target_service.go]: Addressed [28fb195](https://github.com/hashicorp/boundary/pull/2613/commits/28fb195b4b4be8ca8e801ecfb432601a883782d2)

File: internal/daemon/controller/handlers/targets/target_service.go [hunk 18]
```diff
@@ -1585,6 +1630,20 @@ func validateUpdateRequest(req *pbs.UpdateTargetRequest) error {
...
```
[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/daemon/controller/handlers/targets/target_service.go]: Addressed [28fb195](https://github.com/hashicorp/boundary/pull/2613/commits/28fb195b4b4be8ca8e801ecfb432601a883782d2)

[the remark this answers is not in the document]
Comment (hugoamvieira) [reply internal/daemon/controller/handlers/targets/target_service.go]: Not sure if unifying here makes sense because the hosts handlers also have this same code - We decided at the time not to extract because we couldn't find a sensible place to put the unified code.

For me it makes sense to either unify all or not at all - But unifying the entire thing means finding a suitable place for the code

File: internal/session/repository_session.go [hunk 19]
```diff
@@ -34,21 +34,18 @@ func (r *Repository) CreateSession(ctx context.Context, sessionWrapper wrapping.
...
-		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing host set id")
-	}
 	if newSession.AuthTokenId == "" {
 		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing auth token id")
 	}
 	if newSession.ProjectId == "" {
 		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing project id")
 	}
+	if newSession.HostId == "" && newSession.HostSetId == "" && newSession.Endpoint == "" {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing host source and endpoint")
```
[the remark this answers is not in the document]
Comment (ddebko) [reply internal/session/repository_session.go:47]: It's kind of weird because the endpoint can be derived from a host or it can be derived from a target address. So endpoint is always a requirement, but the host and host set are optional.

File: internal/daemon/controller/handlers/targets/target_service.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (ddebko) [reply internal/daemon/controller/handlers/targets/target_service.go]: I agree with Hugo, I'm not sure where we could put the code in the current structure. We might need to add a new package called address or network. But then, this would require moving everything network address related in the repo to this location, in order to maintain consistency. 
It might turn out to be more work and maybe not in the scope of this feature.

File: internal/tests/cli/boundary/target.bats (hunk 9 above)
Comment (ddebko) [reply internal/tests/cli/boundary/target.bats:135]: The bats test does validate this behavior. Because the bats tests are not isolated from each other, the state changes from one test carries on to the next. The following tests does what you suggest:
```
@test "boundary/target: create target with a network address" 
@test "boundary/target: cannot assign an host source to a target with an address"
@test "boundary/target: can assign an host source to a target after deleting address"
@test "boundary/target: cannot assign an address to a target with an host source"
@test "boundary/target: can assign an an address to a target after deleting host source"
```

File: internal/tests/api/targets/target_test.go (hunk 8 above)
Comment (ddebko) [reply internal/tests/api/targets/target_test.go:478]: Addressed [40dfc1f](https://github.com/hashicorp/boundary/pull/2613/commits/40dfc1f9f3a0d16b1934d67583b89a695d7318b7)

File: internal/tests/api/targets/target_test.go (hunk 6 above)
Comment (ddebko) [reply internal/tests/api/targets/target_test.go:418]: The expected status code value is a 400 because grpc-gateway maps FailedPrecondition check errors into bad requests. Here is an example of the output from the cli:
```
Error from controller when performing add-host-sources on target

Error information:
  Kind:                FailedPrecondition
  Message:             target.(Repository).AddTargetHostSources: error creating sets: db.DoTx: target.(Repository).AddTargetHostSources: unable
  to add host sources because a network address is directly assigned to the given target: integrity violation: error #409
  Status:              400
  context:             Error from controller when performing add-host-sources on target
```

File: internal/daemon/controller/handlers/errors.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ddebko) [reply internal/daemon/controller/handlers/errors.go:165]: I would like to follow the pattern or remove it completely to keep things consistent. What do you think?@johanbrandhorst

[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/daemon/controller/handlers/errors.go:165]: Honestly it seems unnecessary to me, but we should keep it consistent and remove it in a separate PR instead.

File: internal/tests/cli/boundary/target.bats (hunk 9 above)
Comment (johanbrandhorst) [reply internal/tests/cli/boundary/target.bats:135]: Cool, thanks!

File: testing/internal/e2e/tests/static/target_address_test.go (hunk 11 above)
Comment (johanbrandhorst) [reply testing/internal/e2e/tests/static/target_address_test.go:189]: OK!

File: internal/tests/api/targets/target_test.go (hunk 6 above)
Comment (johanbrandhorst) [reply internal/tests/api/targets/target_test.go:418]: Ok, fair, but can't we verify that here then? We should be able to parse this error here like we are in the e2e tests, right?

File: internal/session/repository_session.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/session/repository_session.go:47]: Nevermind, this is probably fine as is, I didn't realize you still needed an endpoint when using host sets.

File: internal/db/schema/migrations/oss/postgres_59_01_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/db/schema/migrations/oss/postgres_59_01_test.go:188]: Sounds good 👍🏻.

File: internal/tests/api/targets/target_test.go [hunk 20]
```diff
@@ -379,6 +380,172 @@ func TestList(t *testing.T) {
...
+	require.Empty(t, targetResp.GetItem().Address)
+	version = targetResp.GetItem().Version
+	updateResp, err = tClient.AddHostSources(tc.Context(), targetId, version, []string{hs.Item.Id})
+	require.NoError(t, err)
+	require.NotNil(t, updateResp)
+	require.Empty(t, updateResp.GetItem().Address)
+	require.Equal(t, []string{hs.Item.Id}, updateResp.GetItem().HostSourceIds)
+}
+
+func TestTarget_HostSourceMutualExlcusiveRelationship(t *testing.T) {
```
Comment (johanbrandhorst) [inline internal/tests/api/targets/target_test.go:437]: So close 😂.
```suggestion
func TestTarget_HostSourceMutualExclusiveRelationship(t *testing.T) {
```

File: internal/tests/api/targets/target_test.go (hunk 6 above)
Comment (johanbrandhorst) [reply internal/tests/api/targets/target_test.go:418]: Ah, nevermind, I see this has been added now!

**Review by johanbrandhorst — APPROVED** [review]

File: internal/target/address.go [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+		TargetAddress: &store.TargetAddress{
+			TargetId: targetId,
+			Address:  address,
+		},
+	}
+	return t, nil
+}
+
+// Clone creates a clone of the target address
+func (t *Address) Clone() interface{} {
```
Comment (johanbrandhorst) [inline internal/target/address.go:50]:
```suggestion
func (t *Address) Clone() any {
```

File: internal/tests/api/targets/target_test.go (hunk 20 above)
Comment (ddebko) [reply internal/tests/api/targets/target_test.go]: Fixed.

File: internal/target/address.go (hunk 21 above)
Comment (ddebko) [reply internal/target/address.go]: Fixed.

**Review by ddebko — APPROVED** [review]

**Comment by hugoamvieira** [comment]
Rebasing everything back into the original 10 commits now

File: website/content/docs/concepts/domain-model/index.mdx [hunk 22]
```diff
@@ -51,7 +51,26 @@ The highest-level in Boundary utilizes the concepts of [Roles](/docs/concepts/do
...
+A target exposes two distinct and mutually exclusive configuration paths for its
+representation of a networked service:
+
+- Target with an address
+- Target with host sources
+
+A target with an address represents a target that is set up with a single IP
+address or DNS name directly on the target resource. This mechanism provides
+benefits for new Boundary users and users who don't require Boundary's
+flexibility and scalability, as it's less complex to set up and understand.
```
Comment (tmessi) [inline website/content/docs/concepts/domain-model/index.mdx:65]: I think the phrasing here seems a little odd. Mainly the "don't require Boundary's flexibility and scalability". It isn't "boundary's" flexibility and scalability that is being avoided here, using an Address on a Target is still part of the boundary domain model. Instead it is avoiding the need for Hosts, Host Sets, Host Catalogs. So maybe something like:


```suggestion
address or DNS name directly on the target resource. This mechanism provides
benefits for new Boundary users and users who do not need the flexibility
provided by Hosts, Host Sets, and Host Catalogs.
It is well suited for cases where there are stand-alone hosts
that should not be grouped together and treated as equal from an access control point of view.
```

File: website/content/docs/concepts/domain-model/index.mdx [hunk 23]
```diff
@@ -51,7 +51,26 @@ The highest-level in Boundary utilizes the concepts of [Roles](/docs/concepts/do
...
+Because of its simplicity, using this path on environments where a significant
+number of resources exist is not recommended.
```
Comment (tmessi) [inline website/content/docs/concepts/domain-model/index.mdx:67]:
```suggestion
Because of its simplicity, using targets with an address is not recommended
when there are a either a large number of hosts and/or dynamic hosts.
```

File: website/content/docs/concepts/domain-model/index.mdx [hunk 24]
```diff
@@ -51,7 +51,26 @@ The highest-level in Boundary utilizes the concepts of [Roles](/docs/concepts/do
...
+
+A target with host sources represents a target that has one or more [host
+sets][] associated with it. It trades off the simplicity of directly setting an
+address on a target with Boundary's flexibility and scalability for discovering
+and categorizing networked resources. We recommend this path for setting
+up Boundary in environments where a significant number of resources exist.
```
Comment (tmessi) [inline website/content/docs/concepts/domain-model/index.mdx:73]: This can be reworded similarly to the suggestions above.

File: website/content/docs/concepts/domain-model/index.mdx (hunk 22 above)
Comment (hugoamvieira) [reply website/content/docs/concepts/domain-model/index.mdx:65]: Good point - Makes it much clearer, thanks a lot!

File: website/content/docs/concepts/domain-model/index.mdx (hunk 24 above)
Comment (hugoamvieira) [reply website/content/docs/concepts/domain-model/index.mdx]: Addressed: https://github.com/hashicorp/boundary/pull/2613/commits/cf145792f9ae1161ddc9521953f39ccabd116bab

File: website/content/docs/concepts/domain-model/index.mdx (hunk 23 above)
Comment (hugoamvieira) [reply website/content/docs/concepts/domain-model/index.mdx]: Addressed: https://github.com/hashicorp/boundary/commit/cf145792f9ae1161ddc9521953f39ccabd116bab

File: website/content/docs/concepts/domain-model/index.mdx (hunk 22 above)
Comment (hugoamvieira) [reply website/content/docs/concepts/domain-model/index.mdx]: Addressed: https://github.com/hashicorp/boundary/commit/cf145792f9ae1161ddc9521953f39ccabd116bab

**Review by tmessi — APPROVED** [review]
